core/src/macosX64Main/kotlin/Connection.kt

```diff
@@ -0,0 +1,31 @@
+package com.juul.kable
+
+import com.juul.kable.PeripheralDelegate.DidUpdateValueForCharacteristic
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.sync.Semaphore
+import kotlinx.coroutines.sync.withPermit
+
+internal class Connection(
+    private val delegate: PeripheralDelegate,
+) {
+
+    val characteristicChanges: Flow<DidUpdateValueForCharacteristic> =
+        delegate.characteristicChanges
+
+    // Using Semaphore as Mutex never relinquished lock when multiple concurrent `withLock`s are executed.
```

Comment:
What's the over/under on Mutex being bugged because it accidentally freezes itself somehow?

Reply:
I honestly don't know. I couldn't reproduce on JVM target.
I couldn't find an issue reported for it.

Before I went ahead and reported the issue, I wanted to have a small code snippet reproducing it but that was more work than trying `Semaphore` (and seeing that it just worked). 🤷 

I really should try to find some time to repro and report, but `Semaphore(1)` is currently faster than `Mutex` anyways. See https://kotlinlang.slack.com/archives/C1CFAFJSK/p1605617248151000 for details.

Reply:
I tried to do a little digging into the code, and my only insight is "weird that Mutex.kt is 400 lines but Semaphore.kt is 300 lines." Honestly, it might just be an implementation quality thing. 

Also, Semaphore.kt manages to fit a _beautiful_ ASCII-art comment in those 300 lines:

```kotlin

    /*
       The queue of waiting acquirers is essentially an infinite array based on the list of segments
       (see `SemaphoreSegment`); each segment contains a fixed number of slots. To determine a slot for each enqueue
       and dequeue operation, we increment the corresponding counter at the beginning of the operation
       and use the value before the increment as a slot number. This way, each enqueue-dequeue pair
       works with an individual cell. We use the corresponding segment pointers to find the required ones.

       Here is a state machine for cells. Note that only one `acquire` and at most one `release` operation
       can deal with each cell, and that `release` uses `getAndSet(PERMIT)` to perform transitions for performance reasons
       so that the state `PERMIT` represents different logical states.

         +------+ `acquire` suspends   +------+   `release` tries    +--------+                    // if `cont.tryResume(..)` succeeds, then
         | NULL | -------------------> | cont | -------------------> | PERMIT | (cont RETRIEVED)   // the corresponding `acquire` operation gets
         +------+                      +------+   to resume `cont`   +--------+                    // a permit and the `release` one completes.
            |                             |
            |                             | `acquire` request is cancelled and the continuation is
            | `release` comes             | replaced with a special `CANCEL` token to avoid memory leaks
            | to the slot before          V
            | `acquire` and puts    +-----------+   `release` has    +--------+
            | a permit into the     | CANCELLED | -----------------> | PERMIT | (RElEASE FAILED)
            | slot, waiting for     +-----------+        failed      +--------+
            | `acquire` after
            | that.
            |
            |           `acquire` gets   +-------+
            |        +-----------------> | TAKEN | (ELIMINATION HAPPENED)
            V        |    the permit     +-------+
        +--------+   |
        | PERMIT | -<
        +--------+  |
                    |  `release` has waited a bounded time,   +--------+
                    +---------------------------------------> | BROKEN | (BOTH RELEASE AND ACQUIRE FAILED)
                           but `acquire` has not come         +--------+
    */
```

Reply:
Things always work better when they have ASCII art in their documentation!